### PR TITLE
Minor thread affinity clean-up - Remove redundant UI delay switches in NuGet Project initialization

### LINKED ISSUE #10824: Remove redundant UI delay switches in NuGet Project initialization
There are some redundant UI thread affinity:

For example: 
* https://github.com/NuGet/NuGet.Client/blob/3b1d6b8ab827236ea92fb25fb4e6c29596cf6c47/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/MSBuildNuGetProjectSystemFactory.cs#L34 is unnecessary because all methods using the UI thread are async themselves, so they'll switch if needed.

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/10824

Regression? Last working version:

## Description

See inline comments on the changes for reasoning

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A  Existing tests should catch any issues.

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/MSBuildNuGetProjectSystemFactory.cs [hunk 1]
```diff
@@ -31,8 +31,6 @@ internal static class MSBuildNuGetProjectSystemFactory
 
         public async static Task<VsMSBuildProjectSystem> CreateMSBuildNuGetProjectSystemAsync(IVsProjectAdapter vsProjectAdapter, INuGetProjectContext nuGetProjectContext)
         {
-            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/MSBuildNuGetProjectSystemFactory.cs:34]: Only async methods called here need the UI thread, this method itself doesn't have a UI thread affinity.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs [hunk 2]
```diff
@@ -63,8 +63,6 @@ public sealed class LegacyPackageReferenceProjectProvider : INuGetProjectProvide
         {
             Assumes.Present(vsProjectAdapter);
 
-            await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs:66]: Only async methods called here need the UI thread, this method itself doesn't have a UI thread affinity.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs [hunk 3]
```diff
@@ -92,6 +90,8 @@ public sealed class LegacyPackageReferenceProjectProvider : INuGetProjectProvide
         {
             var componentModel = await _componentModel.GetValueAsync();
 
+            await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs:93]: This method casts to VSProject4 which is a STA bound API. 

The initialization of the IComponentModel is freethreaded, so no need to switch to the UI before that.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/MSBuildNuGetProjectProvider.cs [hunk 4]
```diff
@@ -56,8 +56,6 @@ public MSBuildNuGetProjectProvider(IVsProjectThreadingService threadingService)
             Assumes.Present(vsProjectAdapter);
             Assumes.Present(context);
 
-            await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/MSBuildNuGetProjectProvider.cs:59]: No casting/UI thread affinity in this method, only the async methods called by this method have UI thread affinity.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs [hunk 5]
```diff
@@ -218,29 +218,9 @@ public string Version
 
         #region Getters
 
-        public async Task<string[]> GetProjectTypeGuidsAsync()
+        public Task<string[]> GetProjectTypeGuidsAsync()
         {
-            if (!IsDeferred)
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs:223]: IsDeferred is dead code, not removed by mistake an in earlier clean-up.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceLocator.cs [hunk 6]
```diff
@@ -118,7 +117,7 @@ public static void InitializePackageServiceProvider(IAsyncServiceProvider provid
                 }
             }
 
-            return Package.GetGlobalService(typeof(TService)) as TInterface;
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceLocator.cs]: Dead code. This method requires the UI thread, which defeats the purpose of this method which indicates that it's free threaded.

**Review by zkat — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs [hunk 7]
```diff
@@ -218,29 +218,9 @@ public string Version
...
-                }
-
-                if (!string.IsNullOrEmpty(_projectTypeGuid))
-                {
-                    return new string[] { _projectTypeGuid };
-                }
-
-                return Array.Empty<string>();
-            }
+            return Project.GetProjectTypeGuidsAsync();
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs:223]: discussion starter: https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task

**Review by zivkan — APPROVED** [review]

**Review by erdembayar — APPROVED** [review]

**Pull request merged**